src/helpers/sorobanUtils.ts

```diff
@@ -171,6 +173,125 @@ export const buildTxWithSorobanData = ({
     .build();
 };
 
+export const getTxnToSimulate = (
+  value: SorobanInvokeValue,
+  txnParams: TransactionBuildParams,
+  operation: TxnOperation,
+  networkPassphrase: string,
+): { xdr: string; error: string } => {
+  try {
+    const argsToScVals = getScValsFromArgs(value.args);
+    const builtXdr = buildTxWithSorobanData({
+      params: txnParams,
+      sorobanOp: {
+        ...operation,
+        params: {
+          ...operation.params,
+          contract_id: value.contract_id,
+          function_name: value.function_name,
+          args: argsToScVals,
+          resource_fee: BASE_FEE, // bogus resource fee for simulation purpose
+        },
+      },
+      networkPassphrase,
+    });
+
+    return { xdr: builtXdr.toXDR(), error: "" };
+  } catch (e: any) {
+    return { xdr: "", error: e.message };
+  }
+};
+
+const convertObjectToScVal = (obj: Record<string, any>): xdr.ScVal => {
+  const convertedValue: Record<string, any> = {};
+  const typeHints: Record<string, [string, string]> = {};
+  // obj input example:
+  //  {
+  //    "address": {
+  //      "value": "CDVQVKOY2YSXS2IC7KN6MNASSHPAO7UN2UR2ON4OI2SKMFJNVAMDX6DP",
+  //      "type": "Address"
+  //    },
+  //    "amount": {
+  //      "value": "2",
+  //      "type": "I128"
+  //    },
+  //    "request_type": {
+  //      "value": "4",
+  //      "type": "U32"
+  //    }
+  //  }
+
+  //  for an array of objects, `nativeToScVal` expects the following type for `val`:
+  //  https://stellar.github.io/js-stellar-sdk/global.html#nativeToScVal
+  //  {
+  //     "address": "CDVQVKOY2YSXS2IC7KN6MNASSHPAO7UN2UR2ON4OI2SKMFJNVAMDX6DP",
+  //     "amount": "2",
+  //     "request_type": "2"
+  //  }
+  //  for `type`, it expects the following type:
+  //   {
+  //     "address": [
+  //       "symbol", // matching the key type
+  //       "address" // matching the value type
+  //     ],
+  //     "amount": [
+  //       "symbol", // matching the key type
+  //       "i128" // matching the value type
+  //     ],
+  //     "request_type": [
+  //       "symbol", // matching the key type
+  //       "u32" // matching the value type
+  //     ]
+  //  }
+
+  for (const key in obj) {
+    convertedValue[key] = obj[key].value;
+    // toLowerCase() is needed because the type we save from the label is in uppercase
+    // but nativeToScval expects the type to be in lowercase
+    typeHints[key] = ["symbol", obj[key].type.toLowerCase()];
+  }
+
+  return nativeToScVal(convertedValue, { type: typeHints });
+};
+
+const getScValsFromArgs = (args: SorobanInvokeValue["args"]): xdr.ScVal[] => {
```

Comment:
these are nothing new. we're just moving it to `sorobanUtils`. these got approved in the previous PR #1386